lib/faraday/connection.rb

```diff
@@ -384,11 +388,30 @@ def build_request(method)
       Request.create(method) do |req|
         req.params  = self.params.dup
         req.headers = self.headers.dup
-        req.options = self.options.merge(:proxy => self.proxy)
+        req.options = self.options
         yield(req) if block_given?
       end
     end
 
+    def proxy_allowed?(url)
+      return true if @no_proxy.nil?
+      uri = Utils.URI(url)
+      @no_proxy.none? do |no_proxy_domain|
+        no_proxy_uri = Utils.URI(no_proxy_domain)
+        # assume strings without scheme are of http
+        if no_proxy_uri.class == URI::Generic && !no_proxy_domain.empty?
+          no_proxy_uri = Utils.URI("http://#{no_proxy_domain}")
+        end
+        if no_proxy_uri.port == 80
+          # domain matching is fine
+          uri.host.downcase.end_with?(no_proxy_uri.host)
+        else
+          # need to match ports exactly
+          "#{uri.host.downcase}:#{uri.port}".end_with?("#{no_proxy_uri.host}:#{no_proxy_uri.port}")
+        end
```

Comment:
`uri.port` will default to 80. It would be simpler and remove the need to insert the hosts and ports into strings by just comparing the hosts and ports separately. 

So replace the `if no_proxy_uri.port == 80` statement (lines 405-411) with:

`uri.host.downcase.end_with?(no_proxy_uri.host) && uri.port == no_proxy_uri.port
`